parquet-column/src/main/java/org/apache/parquet/column/values/bloomfilter/BlockSplitBloomFilter.java

```diff
@@ -0,0 +1,294 @@
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.parquet.column.values.bloomfilter;
+import com.google.common.hash.HashFunction;
+import com.google.common.hash.Hashing;
+import org.apache.parquet.Preconditions;
+import org.apache.parquet.bytes.BytesUtils;
+import org.apache.parquet.io.api.Binary;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.nio.IntBuffer;
+
+/*
+ * This Bloom filter is implemented using block-based Bloom filter algorithm from Putze et al.'s
+ * "Cache-, Hash- and Space-Efficient Bloom filters". The basic idea is to hash the item to a tiny
+ * Bloom filter which size fit a single cache line or smaller. This implementation sets 8 bits in
+ * each tiny Bloom filter. Each tiny Bloom filter is 32 bytes to take advantage of 32-byte SIMD
+ * instruction.
+ */
+public class BlockSplitBloomFilter implements BloomFilter {
+  // Bytes in a tiny Bloom filter block.
+  private static final int BYTES_PER_BLOCK = 32;
+
+  // Default seed for the hash function. It comes from System.nanoTime().
+  private static final int DEFAULT_SEED = 1361930890;
+
+  // Minimum Bloom filter size, set to the size of a tiny Bloom filter block
+  public static final int MINIMUM_BYTES = 32;
+
+  // Maximum Bloom filter size, set to the default HDFS block size for upper boundary check
+  // This should be re-consider when implementing write side logic.
+  public static final int MAXIMUM_BYTES = 128 * 1024 * 1024;
+
+  // The number of bits to set in a tiny Bloom filter
+  private static final int BITS_SET_PER_BLOCK = 8;
+
+  // The metadata in the header of a serialized Bloom filter is three four-byte values: the number of bytes,
+  // the filter algorithm, and the hash algorithm.
+  public static final int HEADER_SIZE = 12;
+
+  // The default false positive probability value
+  public static final double DEFAULT_FPP = 0.01;
+
+  // Hash strategy used in this Bloom filter.
+  public final HashStrategy hashStrategy;
+
+  // The underlying byte array for Bloom filter bitset.
+  private byte[] bitset;
+
+  // A integer array buffer of underlying bitset to help setting bits.
+  private IntBuffer intBuffer;
```

Comment:
Jim's comments: [https://github.com/apache/parquet-mr/pull/521#discussion_r228761288](url)

Just get some time for this. JMeter seems not suit for showing the memory layout for java heap. I wrote a example application and use jmap to dump the heap.   The code snippet is like: 
```java
        byte [] bitset = new byte[1024*1024];
        IntBuffer intBuffer = ByteBuffer.wrap(bitset).asIntBuffer();
        ByteBuffer byteBuffer = ByteBuffer.allocate(1024*1024);
        byte[] bitset2 = new byte[1024*1024];
```

The Eden space in heap dumps after every sentence are: 

Eden Space:                                                                                                                                                                                                 
   capacity = 66060288 (63.0MB)                                                                                                                                                                             
   used     = 3691032 (3.5200424194335938MB)                                                                                                                                                                
   free     = 62369256 (59.479957580566406MB)                                                                                                                                                               
   5.587368919735863% used    

Eden Space:
   capacity = 66060288 (63.0MB)
   used     = 3691032 (3.5200424194335938MB)
   free     = 62369256 (59.479957580566406MB)
   5.587368919735863% used

Eden Space:
   capacity = 66060288 (63.0MB)
   used     = 4739624 (4.520057678222656MB)
   free     = 61320664 (58.479942321777344MB)
   7.17469472733755% used

Eden Space:
   capacity = 66060288 (63.0MB)
   used     = 5788216 (5.520072937011719MB)
   free     = 60272072 (57.47992706298828MB)
   8.762020534939236% used

The Eden space does not increase when we call the wrap API of ByteBuffer.